### DISCUSSION TITLE
Colorsphere - an interactive 3D color picker for your Twinkly lights

### DISCUSSION THREAD

**Discussion by Anders-Holst** [Anders-Holst/xled_plus#5] [Effects]
There is now an interactive color picker included in xled_plus. It uses an experimental 3D-sphere representation of the entire color body, where one pole is black, the other pole is white, and the color circle is around the equator. If you follow a meridian from the black pole, the color will gradually increase in strength to its maximum brilliance and then seamlessly continue to become brighter all the way to white.

You will see the sphere from one side at a time. To rotate it, grab the surface and drag. When you move the mouse move over the sphere, your leds will light up with the corresponding color. If you click on the sphere, that color will be uploded (as a still movie) to your leds and remain on the leds when you leave the sphere.

Less saturated colors are found inside the sphere. To reach inside, use the control-key together with the scroll-wheel.

To get started with the color-picker, you need to install both the latest (or at least todays, Dec 7) version of xled_plus, and the colorsphere package (which in turns requires matplotlib and numpy):

`pip install -U xled_plus`
`pip install colorsphere`

Then make sure your leds are connected, and start the color picker as:

`python -m xled_plus.xled_colorpicker`

If you want to print the RGB coordinates of the selected color instead of uploading a movie, you can start it with:

`python -m xled_plus.xled_colorpicker rgb noset`

Press "q" in the window to kill it.

Good luck!

**Comment by gonzotek** [comment]
You've been doing some amazing work! Cheers!  I've been doing my best to 'follow along at home'. I'm running xled_plus in a venv on a RPi4 running debian buster (old, I know!).  This machine runs my Home Assistant instance along with a number of other services for my house so switching to something newer isn't easy.  I was able to get xled_plus running pretty easily. I had to build pyzmq using the "--no-binary" pip directive, and colorsphere needed the same for both numpy and matplotlib...I didn't log issues on these because my environment is pretty heavily customized (compiled python 3.9 from source to accommodate Home Assistant requirements) and I don't think it's an issue most people would encounter.  

However, as this machine is headless, is there any hope of accessing colorsphere remotely?  I don't have another good python3 environment at home (work laptop with restrictions on what I can install and wife's chromebook).  After compiling numpy and matplotlib, it appears to run fine from an ssh terminal session, but of course it just sits there since whatever it's doing graphically is happening on the host.  I don't have X.org or similar running on it, up to this point, I didn't need any UI for the other home server stuff I use.  I'll understand if it's a non-starter!

**Reply by Anders-Holst** [reply]
I can understand your situation  - I think many run RP for IoT stuff, and many have played with their systems so it is difficult to upgrade. This is why I have not made colorsphere a requirement for xled_plus, because it could mean trouble for some.

I think there should be some different ways to run it remotely, depending on system setup. (However, this is not my area really, so I may be mistaken. If anyone knows better ways, please help us out.)

First, X11 can run on a remote display. So if you would have had X11 on RP, and another computer with an X11 server, you can just set the DISPLAY environment variable to the other computer. If I remember right there is even an X11 server in cygwin, so you should be able to get the colorsphere displaying even on a windows machine. 

However, for working with X11 I think that matplotlib wants a backend, for example tkinter, which I had troubles installing on an older machine. But I am very uncertain what the options are here.

A second way to run remotely is through a jupyter notebook (or jupyterlab or smilar). I am not sure what it takes to install a jupyter server on you RP. When you have started a notebook, you have to give the magic command `%matplotlib inline` at the prompt (and a different one in jupyterlab which I don't remember right now), for graphics to come up. Then you should be able to start the colorsphere like:
```
host = discover().ip_address
ctr = HighControlInterface(host)
XledColorPicker(ctr).launch(printhsl=True)
```
(Maybe the frame looks uglier than it was designed to, and the scrollwheel doesn't work well since those events are stolen by the browser, but otherwise I think it should work.)

Finally, if you fail to connect the colorsphere to xled, you can run it standalone on another computer:
`python -m colorsphere`
It will not control the leds, but it will print the color coordinates where you click, so you can cut those coordinates into the xled effect, as in todays (December 7) daily effect #2 .

Hope you manage to get something of this to work!

**Reply by gonzotek** [reply]
Thanks for the thoughtful reply!  After work this evening and next (I'm in the US, EST timezone), I'll play with this some more.  I remembered that I have a spare rpi3 that I could install a fresh os with a full X11/gnome enviro, so I think it'll be easiest to setup xled/colorsphere on that, at least for the rest of this holiday season. Then I'll be able to vnc into it to play with colorsphere.  Loving all of the effects so far. I've gotten as far as adding buttons for them into the Home Assistant ui, so my wife can easily use them as well.

**Comment by gonzotek** [comment]
I went with the spare device in the end.  I don't have the time right now to get in to messing with my main home controller.  colorsphere installed and ran wonderfully. Thanks!
